src/lsm/README.md

```diff
@@ -0,0 +1,215 @@
+  - The mutable table's size is allocated to accommodate a full bar of updates.
+- The [immutable table](table_immutable.zig) is an in-memory table.
+  - Each tree has a single immutable table.
+  - The mutable table's contents are periodically moved to the immutable table,
+    where they are stored while being flushed to level `0`.
+- Level `0` … level `config.lsm_levels - 1` each contain an exponentially increasing number of
+  on-disk tables.
+  - Each tree has as many as `config.lsm_growth_factor ^ (level + 1)` tables per level.
+    (`config.lsm_growth_factor` is typically 8).
+  - Within a given level and snapshot, the tables' key ranges are [disjoint](manifest_level.zig).
+
+## Compaction
+
+Tree compaction runs to the sound of music!
+
+Compacting LSM trees involves merging and moving tables into the next levels as needed.
+To avoid write amplification stalls and bound latency, compaction is done incrementally.
+
+A full compaction phase is denoted as a bar or measure, using terms from music notation.
+Each bar consists of `lsm_batch_multiple` beats or "compaction ticks" of work.
+A compaction tick executes asynchronously immediately after every commit, with
+`beat = commit.op % lsm_batch_multiple`.
+
+A bar is split in half according to the "first" beat and "middle" beat.
+The first half of the bar compacts even levels while the latter compacts odd levels.
+Mutable table changes are sorted and compacted into the immutable table.
+The immutable table is compacted into level 0 during the odd level half of the bar.
+
+At any given point, there are at most `levels/2` compactions running concurrently.
+The source level is denoted as `level_a` and the target level as `level_b`.
+The last level in the LSM tree has no target level so it is never a source level.
+Each compaction compacts a [single table](#table-selection) from `level_a` into all tables in
+`level_b` which intersect the `level_a` table's key range.
+
+Invariants:
+* At the end of every beat, there is space in mutable table for the next beat.
+* The manifest is compacted at the end of every beat.
+* The compactions' output tables are not [visible](#snapshots-and-compaction) until the compaction has finished.
+
+1. First half-bar, first beat ("first beat"):
+    * Assert no compactions are currently running.
+    * Allow the per-level table limits to overflow if needed (for example, if we may compact a table
+      from level `A` to level `B`, where level `B` is already full).
+    * Start compactions from even levels that have reached their table limit.
+
+2. First half-bar, last beat:
+    * Finish ticking any incomplete even-level compactions.
+    * Assert on callback completion that all compactions are complete.
+
+3. Second half-bar, first beat ("middle beat"):
+    * Assert no compactions are currently running.
+    * Start odd level compactions if there are any tables to compact.
+    * Start compactions from odd levels that have reached their table limit.
```

Comment:
Are these two points distinct?

Reply:
No, it looks like I reworded it but left the old version in :facepalm:.

Reply:
The second is my (attempted) improvement, to be more explicit about which is the source level, and also that "any tables to compact" refers to.